### PR TITLE
Replace sympy.Mod with custom Mod object

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96270

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/fx/interpreter.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def run(self, *args, initial_env : Optional[Dict[Node, Any]] = None, enable_io_p
             try:
                 self.env[node] = self.run_node(node)
             except Exception as e:
-                if self.extra_traceback:
+                if hasattr(self, "extra_traceback") and self.extra_traceback:
```
Comment (ezyang) [inline torch/fx/interpreter.py:139]: 🤔

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -648,6 +648,53 @@ def check_supported_type(x):
...
+    @classmethod
+    def eval(cls, base, modulus):
+        if base == 0 or modulus == 1:
+            return sympy.Integer(0)
+
+        if (
+            isinstance(base, sympy.Integer)
+            and isinstance(modulus, sympy.Integer)
+        ):
+            return base % modulus
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:674]: Upstream mod has a number of seemingly non-objectionable other simplifications, which may be worth taking. See https://github.com/sympy/sympy/blob/master/sympy/core/mod.py

At a brief glance, here are rules we might like:

* Mod by two can use even/odd assumptions
* Simplifying nested Mod

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -648,6 +648,53 @@ def check_supported_type(x):
...
+
+        if isinstance(base, sympy.Add):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:676]: sympy has a comment like

```
        # (x + y + 2) % x -> Mod(y + 2, x)
```

which I thought was pretty useful for understanding what their Add rule is simplifying for

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -648,6 +648,53 @@ def check_supported_type(x):
...
+            new_terms = []
+            all_positive = True
+            for term in base.args:
+                if sympy.gcd(term, modulus) != modulus:
+                    if (isinstance(term, sympy.Integer) and term < 0) or (
+                        isinstance(term, sympy.Mul)
+                        and isinstance(term.args[0], sympy.Integer)
+                        and term.args[0] < 0
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:684]: This condition looks wrong. Why can't we use sympy assumptions to determine if the term is negative or not?

**Review by ezyang — CHANGES_REQUESTED** [review]
Not convinced by the Triton workaround code

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:684]: This was just mimicked from the existing ModularIndexing code. cc: @ngimel

Comment (ngimel) [reply torch/fx/experimental/symbolic_shapes.py:684]: I'm not against using sympy assymptions, I think when I just tried to put comparison in ModularIndexing it errored out, but I don't remember. 
This exemption shouldn't be necessary really, in ModularIndexing it exists only to work around triton bug. (e.g. with valid simplifications (x - y) % x will turn in to -y % x and iirc triton has problems with negative FloorDivs or modulos, but we probably don't need to care about it in symbolic_shapes

**Pull request closed without merging**